nautobot/extras/forms.py

```diff
@@ -259,41 +259,50 @@ def _save_relationships(self):
             if hasattr(self.cleaned_data[field_name], "__iter__"):
+                # One-to-many or many-to-many association
                 target_peer_ids = [item.pk for item in self.cleaned_data[field_name]]
             elif self.cleaned_data[field_name]:
+                # Many-to-one or one-to-one association
                 target_peer_ids = [self.cleaned_data[field_name].pk]
             else:
-                continue
+                # Unset/delete case
+                target_peer_ids = []
 
-            # Delete all existing CRA that are not in cleaned_data/target_peer_ids list
-            # Remove from target_peer_ids all peers that already exist
-            for cra in existing_cras:
-                found_peer = False
+            # Create/delete RelationshipAssociations as needed to match the target_peer_ids list
+
+            # First, for each existing association, if it's one that's already in target_peer_ids,
+            # we can discard it from target_peer_ids (no update needed to this association).
+            # Conversely, if it's *not* in target_peer_ids, we should delete it.
+            for association in existing_associations:
                 for peer_id in target_peer_ids:
-                    if peer_id == getattr(cra, f"{peer_side}_id"):
-                        found_peer = peer_id
-                if not found_peer:
-                    cra.delete()
+                    if peer_id == getattr(association, f"{peer_side}_id"):
+                        # This association already exists, so we can ignore it
+                        target_peer_ids.remove(peer_id)
+                        break
                 else:
-                    target_peer_ids.remove(found_peer)
+                    # This association is not in target_peer_ids, so delete it
+                    association.delete()
 
-            for cra_peer_id in target_peer_ids:
+            # Anything remaining in target_peer_ids now does not exist yet and needs to be created.
+            for peer_id in target_peer_ids:
                 relationship = self.fields[field_name].model
-                cra = RelationshipAssociation(
+                association = RelationshipAssociation(
                     relationship=relationship,
+                    **{
+                        f"{side}_type": self.obj_type,
+                        f"{side}_id": self.instance.pk,
+                        f"{peer_side}_type": getattr(relationship, f"{peer_side}_type"),
+                        f"{peer_side}_id": peer_id,
+                    },
                 )
-                setattr(cra, f"{side}_id", self.instance.pk)
-                setattr(cra, f"{side}_type", self.obj_type)
-                setattr(cra, f"{peer_side}_id", cra_peer_id)
-                setattr(cra, f"{peer_side}_type", getattr(relationship, f"{peer_side}_type"))
 
                 # FIXME Run Clean
-                cra.save()
+                association.save()
```

Comment:
That `FIXME` just above is staring me in the face. It seems like this is the right time to add the call to `clean()` on the model? Since `RelationshipAssociation` doesn't have its own form, it seems to me that this means the validation in `RelationshipAssociation.clean()` isn't actually being performed at all?

Reply:
I'd left this FIXME alone since I'm not sure 1) why it wasn't done in the first place and 2) what the implications are of adding it. In particular it occurs to me that this method (`_save_relationships`) is being called from the form `save()` method, so it's arguably too late to be calling the model `clean()` (which would normally be called from the ModelForm's `clean()` method.

Reply:
The lack of calling `RelationshipAssociation.clean()` during form cleaning is probably a root cause of #332. So if it's OK with you, I'll defer this until fixing that issue?